code/datums/components/transforming.dm

```diff
@@ -0,0 +1,241 @@
+/*
+ * Transforming weapon component. For weapons that swap between states.
+ * For example: Energy swords, cleaving saws, switch blades.
+ *
+ * Used to easily make an item that can be attack_self'd to gain force or change mode.
+ *
+ * Only values passed on initialize will update when the item is activated (except the icon_state).
+ * The icon_state of the item will swap between "[icon_state]" and "[icon_state]_on".
+ */
+/datum/component/transforming
+	/// Whether the weapon is transformed
+	var/active = FALSE
+	/// Cooldown on transforming this item back and forth
+	var/transform_cooldown_time
+	/// Force of the weapon when active
+	var/force_on = 0
+	/// Throwforce of the weapon when active
+	var/throwforce_on = 0
+	/// Weight class of the weapon when active
+	var/w_class_on = WEIGHT_CLASS_BULKY
+	/// The sharpness of the weapon when active
+	var/sharpness_on = NONE
+	/// Hitsound played when active
+	var/hitsound_on = 'sound/weapons/blade1.ogg'
+	/// List of continuous attack verbs used when the weapon is enabled
+	var/list/attack_verb_continuous_on = list("attacks", "slashes", "stabs", "slices", "tears", "lacerates", "rips", "dices", "cuts")
+	/// List of simple attack verbs used when the weapon is enabled
+	var/list/attack_verb_simple_on = list("attack", "slash", "stab", "slice", "tear", "lacerate", "rip", "dice", "cut")
+	/// Whether clumsy people need to succeed an RNG check to turn it on without hurting themselves
+	var/clumsy_check = TRUE
+	/// If we get sharpened with a whetstone, save the bonus here for later use if we un/redeploy
+	var/sharpened_bonus
+	/// Callback to be invoked whenever the weapon is transformed.
+	var/datum/callback/on_transform_callback
+	/// Cooldown in between transforms
+	COOLDOWN_DECLARE(transform_cooldown)
+
+/datum/component/transforming/Initialize(
+		start_transformed = FALSE,
+		transform_cooldown_time,
+		force_on,
+		throwforce_on,
+		sharpness_on,
+		hitsound_on,
+		w_class_on,
+		clumsy_check = TRUE,
+		list/attack_verb_continuous_on,
+		list/attack_verb_simple_on,
+		datum/callback/on_transform_callback,
+		)
+
+	if(!isitem(parent))
+		return COMPONENT_INCOMPATIBLE
+
+	src.transform_cooldown_time = transform_cooldown_time
+	src.force_on = force_on
+	src.throwforce_on = throwforce_on
+	src.sharpness_on = sharpness_on
+	src.hitsound_on = hitsound_on
+	src.w_class_on = w_class_on
+	src.clumsy_check = clumsy_check
```

Comment:
you're going to be replacing src values with null without the default values being sent.